hilti/runtime/CMakeLists.txt

```diff
@@ -74,19 +74,59 @@ foreach ( lib hilti-rt hilti-rt-debug )
-target_compile_options(hilti-rt-objects     PRIVATE "-g;-O3;-DNDEBUG;-Wall")
-target_compile_definitions(hilti-rt-objects PRIVATE "HILTI_RT_BUILD_TYPE_RELEASE")
-
-# Build hilti-rt-debug with debug flags.
-string(REPLACE " " ";" cxx_flags_debug ${CMAKE_CXX_FLAGS_DEBUG})
-target_compile_options(hilti-rt-debug-objects     PRIVATE ${cxx_flags_debug})
-target_compile_options(hilti-rt-debug-objects     PRIVATE "-UNDEBUG;-O0;-Wall")
-target_compile_definitions(hilti-rt-debug-objects PRIVATE "HILTI_RT_BUILD_TYPE_DEBUG")
+    if ("${lib}" MATCHES "debug$")
+        string(REPLACE " " ";" cxx_flags_debug ${CMAKE_CXX_FLAGS_DEBUG})
+        target_compile_options(hilti-rt-debug-objects     PRIVATE ${cxx_flags_debug})
+        target_compile_options(hilti-rt-debug-objects     PRIVATE "-UNDEBUG;-O0;-Wall")
+        target_compile_definitions(hilti-rt-debug-objects PRIVATE "HILTI_RT_BUILD_TYPE_DEBUG")
+    else ()
+        string(REPLACE " " ";" cxx_flags_release ${CMAKE_CXX_FLAGS_RELEASE})
+        target_compile_options(hilti-rt-objects     PRIVATE ${cxx_flags_release})
+        target_compile_options(hilti-rt-objects     PRIVATE "-g;-O3;-DNDEBUG;-Wall")
+        target_compile_definitions(hilti-rt-objects PRIVATE "HILTI_RT_BUILD_TYPE_RELEASE")
+    endif ()
+
+    # Precompile libhilti for use in JIT during development. Users should use `scripts/precompile-headers.sh`.
+    # Create a precompiled header for libhilti.h.
+    #
+    # We generate the precompiled header by hand and only use it during JIT,
+    # but e.g., not to during compilation of Spicy itself. This gives us the
+    # benefits of JIT without e.g., making it harder for ccache to work during
+    # development. It also allows us to punt on some trickier cleanups of
+    # header files.
+    #
+    # TODO(bbannier): We currently only provide a precompiled header in the
+    # build directory but not one during install or packaging. Especially
+    # packaging appears to be tricky and precompiled headers seem not in
+    # general movable after creation. We should figure out a way to give users
+    # the same speed-up during JIT we get during development from using a
+    # precompiled header.
+    set(debug_opt "")
+    if ("${lib}" MATCHES "debug$")
+        set(debug_opt "_debug")
+    endif ()
+
+    get_target_property(PCH_INCLUDES ${lib}-objects INCLUDE_DIRECTORIES)
+    list(TRANSFORM PCH_INCLUDES PREPEND "-I")
+    get_target_property(PCH_OPTIONS ${lib}-objects COMPILE_OPTIONS)
+    message("OPTIONS " ${PCH_OPTIONS_})
+    add_custom_command(
+        OUTPUT ${AUTOGEN_H}/libhilti${debug_opt}.h.pch
+        COMMAND ${CMAKE_CXX_COMPILER} --std=c++${CMAKE_CXX_STANDARD} -x c++-header
+            "${CMAKE_CURRENT_SOURCE_DIR}/include/libhilti.h"
+            -Winvalid-pch ${PCH_OPTIONS} ${PCH_INCLUDES}
+            -o ${AUTOGEN_H}/precompiled_libhilti${debug_opt}.h.pch
+        DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/include/libhilti.h)
```

Comment:
Could we reuse the shell script for this if we gave it an option to specify where to put the precompiled headers? Would be nice to have the logic at only one location.

Reply:
This is a little tricky to do as we'd have to know the build directory from the script. We could either generate this script during the cmake configuration phase or much simpler, leave a comment here and in the cmake setup to keep the flags in sync.

I'd prefer the comment. WDYT?

Reply:
I think I'm missing something: can't we just pass the directory in through a command line argument?

Reply:
Sorry, I mixed up two things: We'd need the path to `hilti-config`, not to the build dir (the former can provide us the latter).

I'll add behavior allowing to inject the location of `hilti-config`.

Reply:
Also changed the cmake setup to use the script.